### PR TITLE
Change/remove a branch of an open issue

### LINKED ISSUE #3620: Modifying branch in issues
From the forums: https://discourse.gitea.io/t/associating-existing-issue-with-a-branch/218/2?u=howl

#780 added the ability to link an issue to a certain branch. However, the branch can't be modified or removed afterwards, so that should be added.

### PR DESCRIPTION
**Pull request by vedranMv**
Fix #3620  This PL addressed inability to change a branch of an open issue

b84622d Follows the pattern of changing a title of an issue (in the backend) and coupled with frontend changes allows changing the branch of an open issue. However, no comments are currently posted to the issue when a branch is changed. I'm still not sure about the best way to implement that but it looks like it might require changes in DB (add old_ref and new_ref columns to 'comment' table).

Perhaps it would also make sense to only merge this part for now, and then follow up with a new PL for comments :)

### CODE REVIEW

**Comment by vedranMv** [comment]
668d3c0  Merged master into branch (I messed up commit message)

**Comment by lunny** [comment]
@vedranMv Could you give any screenshot?

**Comment by vedranMv** [comment]
Sure
branch-select

File: models/issue.go
Comment (guillep2k) [inline models/issue.go]: Why is this needed here?

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: How can the user remove the target branch?

File: templates/repo/issue/branch_selector_field.tmpl
Comment (guillep2k) [inline templates/repo/issue/branch_selector_field.tmpl]:
```suggestion
<input id="editing_mode" name="edit_mode" type="hidden" value="{{(or .IsIssueWriter .IsIssuePoster)}}">
```
This will make it consistent with the backend. However, I'm not sure what the permission requirements should be for changing the issue branch; write issue permissions, merge permissions?

File: routers/repo/issue.go
Comment (vedranMv) [reply routers/repo/issue.go]: So far it can't, however, that's the same problem as when you try to create an issue - once selected, the branch can't be removed short of refreshing the page. I'll look into it 

And thanks for all the feedback, it's pretty useful :)

File: templates/repo/issue/branch_selector_field.tmpl
Comment (vedranMv) [reply templates/repo/issue/branch_selector_field.tmpl]: My idea was that only write access/assignee should be allowed to change branch. In my opinion, creator should not be allowed to change it unless it fits the first criteria (or even assign a branch to an issue in the first place). I would base this on the assumption that an issue can be created by anyone, but only those who can solve it should be able to link it to a branch.

But then I was confused about IsIssueWriter here. Comment in the code says:

> returns true if isPull is true and user could write to pull requests and returns true if isPull is false and user could write to issues

The way I interpreted is that if it can create issues, which any user on a public repo can, it could change a branch. In practice, seems that it needs write access which is what I would prefer here.

File: models/issue.go
Comment (vedranMv) [reply models/issue.go]: I implemented this feature following a pattern of updating a title of an issue. ``updateIssueCols`` is needed to update 'ref' field in DB and save branch change while loadRepo retrieves this info from DB to make sure changes were indeed implemented which through API is returned to front-end (JS).

But writing this now, I realize that returning this info to JS might be unnecessary since the page is anyway reloaded once POST is completed.

File: templates/repo/issue/branch_selector_field.tmpl
Comment (guillep2k) [reply templates/repo/issue/branch_selector_field.tmpl]: It's confusing indeed, but it's not what it seems:

**Issue read permissions**: Can read issues, can comment on any issue, can create issues, can delete/modify/close own issues.
**Issue write permissions**: All of the above plus can delete/modify/close other people's issues.

Hope this helps. 😄

File: models/issue.go
Comment (guillep2k) [reply models/issue.go]: There's a tendency to avoid reloading the page if not required (so a half-typed comment will not be lost), but you don't need to concern about that in this PR unless you want to dig deeper.

Having said that, `loadRepo()` should not be called in a `models` function unless it's required for that function. If you happend to need that, you can call `LoadRepo()` from `services` or `routers` as you see fit.

Comment (vedranMv) [reply models/issue.go]: Oh I see, I'll remove it for this PR. But after loading the repo, I'd like to call ``createComment`` which I'm testing in another branch. And there, as far as I understand, I'll need it
image

File: templates/repo/issue/branch_selector_field.tmpl
Comment (vedranMv) [reply templates/repo/issue/branch_selector_field.tmpl]: It does indeed, thanks ^^

**Comment by vedranMv** [comment]
20f10af addresses all but one concern, deleting a branch. 
I can't think of a simple solution as it involves completely new front-end element :)

**Comment by guillep2k** [comment]
> 
> 
> [20f10af](https://github.com/go-gitea/gitea/commit/20f10af3cf442d4f16f040f82b330fd3a9a8ed8a) addresses all but one concern, deleting a branch.
> I can't think of a simple solution as it involves completely new front-end element :)

Can a ficticious blank (" ") branch name added at the end of the list as a placeholder?

**Comment by vedranMv** [comment]
That's why I like reviews :D I have a tendency to overcomplicate sometimes. I'll fix it later today

**Comment by vedranMv** [comment]
How about 8b68d5d  ? There's now a button which appears only when issue has a ref assigned to it? :)
s1
s2

The drawback of this button is that page now has to be refreshed when selecting a branch otherwise the branch list doesn't get updated and you either can't see the 'clear ref' button after the branch has been selected, or it stays there after you've cleared a branch.

**Review by guillep2k — APPROVED** [review]
This works like a treat!

You may want to update the title of this PR though, to mention the possibility of branch removal. 😉

**Comment by vedranMv** [comment]
fbf7ad9 fixes a bug with branch name not shown when creating an issue. POST is not executed and there's nothing to refresh the field with branch name

branch-issue

Should I also merge master into this branch to resolve JS conflicts?

**Comment by guillep2k** [comment]
> 
> 
> Should I also merge master into this branch to resolve JS conflicts?

Yes. I think you need to merge, recreate index.js/index.js.map, git add, commit and push. Maybe a rebase is possible as well. We're working on alternatives to avoid this kind of conflict, but for the moment it is what it is. 😄

**Comment by vedranMv** [comment]
@guillep2k Any idea why the build is failing? I've changed nothing from my side in the last two commits but both builds failed on a different database test.

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k Any idea why the build is failing? I've changed nothing from my side in the last two commits but both builds failed on a different database test.

Yes, I have an idea and don't worry, it's not your fault. We're having intermittent problems with some of our tests and we are working hard trying to fix them. On top of that, on occasion our CI can have a glitch and tests could fail randomly.

But we are aware of this situation, so we don't penalize PRs because of this (we check the nature of the failure when this happens). We're pretty used to it. 😒

Please be patient about your PR. It will have its second review eventually. We're a little short-handed at the moment. 😁

**Comment by vedranMv** [comment]
No problem, I'm not in a rush 😄 
It was more like 'is there anything I can do to help it'. But thanks for the explanation, I understand

**Comment by lunny** [comment]
@vedranMv @guillep2k And then when a PR merged and it mentioned will close some issue but that issue has a same or different branch as that PR's target branch. How should we handle that situation?

**Comment by saitho** [comment]
Related to #6488

**Comment by guillep2k** [comment]
> 
> 
> @vedranMv @guillep2k And then when a PR merged and it mentioned will close some issue but that issue has a same or different branch as that PR's target branch. How should we handle that situation?

@lunny This is a corrective action from the part of the user, so maybe it's up to the user to check whether the context of the change requires some revision. I see two distinct scenarios:

* The info in the branch field was wrong, but other references across the system are correct (e.g. comments or commits in PRs mention the issue correctly). The user will change the field so those PRs will close the issue properly, as otherwise they wouldn't.
* The info in the branch field is changed for other reasons, invalidating or _activating_ mentions on PRs.

The correctness of these situations is strongly contextual; Gitea has no way of knowing the users' motives and there may be valid reasons to change the value without any modifications to mentions in PRs.

Since the user changing the issue branch already has the power to close/reopen the issue themselves, I don't see any security issues about this either.

IMHO this doesn't need to be addressed by Gitea. The user changing branches already has the required information (in the form of action comments) to make an informed decision about the effect of the change, although that could be improved.

If you're still concerned about this, I suppose we could have a notification issued to the users that created those references.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: @vedranMv This was an oversight from my part. Although the UI will not let the user change the branch on pull requests, please add a check here too so PRs can't be changed by using the API directly.

Also, I think we should also check that the issue is not closed (that check should also be added to the UI).

**Review by guillep2k — CHANGES_REQUESTED** [review]

File: routers/repo/issue.go
Comment (lunny) [reply routers/repo/issue.go]: If user try to change a branch of a PR, we should return an error.

Comment (vedranMv) [reply routers/repo/issue.go]: Sorry for the delay. Something like ~~2bd8e64~~ b913c6e  ?
edit:brainfart :D

**Review by guillep2k — APPROVED** [review]

File: services/issue/issue.go
Comment (lunny) [inline services/issue/issue.go]: We need a new notification - change branch

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by vedranMv** [comment]
Sorry for the delay.
I have to admit I don't really understand the workflow for notifications here. I tried following the logic of when a title/label/assignee/etc is changed, but these notifications don't seem to work (i.e. when changing a title, I can't get a notification on a different user who is subscribed to an open issue, has commented on it and has write access to repo).
Maybe somebody can point what's missing from d7cd33e ? :)

File: web_src/js/index.js
Comment (lunny) [inline web_src/js/index.js:127]: I think we could not reload it and just change the value.

**Review by lunny — CHANGES_REQUESTED** [review]

File: web_src/js/index.js
Comment (lafriks) [reply web_src/js/index.js:127]: This could be done in other PR along with assignee, label and milestone change

**Comment by lafriks** [comment]
Please resolve conflicts

**Comment by vedranMv** [comment]
> 
> 
> Please resolve conflicts

Done. But when I try to compile and run the branch I get runtime exception related to translations. Same exception if I compile _master_ branch. Any idea what I messed up? Here is the exception:

```
admin001@homeserver:~/git/gitea$ TAGS="bindata sqlite sqlite_unlock_notify" make build
admin001@homeserver:~/git/gitea$ ./gitea web --port=5000
2020/05/17 16:01:58 cmd/web.go:107:runWeb() [I] Starting Gitea on PID: 2154
2020/05/17 16:01:58 ...dules/setting/git.go:93:newGit() [I] Git Version: 2.17.1
2020/05/17 16:01:58 routers/init.go:119:GlobalInit() [T] AppPath: /home/admin001/git/gitea/gitea
2020/05/17 16:01:58 routers/init.go:120:GlobalInit() [T] AppWorkPath: /home/admin001/git/gitea
2020/05/17 16:01:58 routers/init.go:121:GlobalInit() [T] Custom path: /home/admin001/git/gitea/custom
2020/05/17 16:01:58 routers/init.go:122:GlobalInit() [T] Log path: /home/admin001/git/gitea/log
panic: fail to set message file(es-ES): key-value delimiter not found: Marcar conversación sin resolver


goroutine 1 [running]:
gitea.com/macaron/i18n.initLocales(0xc00020a755, 0x0, 0x2224f24, 0xb, 0xc000d140c0, 0x2293b69, 0x12, 0xc000134420, 0x16, 0x16, ...)
        /home/admin001/git/gitea/vendor/gitea.com/macaron/i18n/i18n.go:57 +0x689
gitea.com/macaron/i18n.I18n(0xc0024b9500, 0x1, 0x1, 0x10, 0xc00348c8b0)
        /home/admin001/git/gitea/vendor/gitea.com/macaron/i18n/i18n.go:160 +0xed
code.gitea.io/gitea/routers.InitLocales()
        /home/admin001/git/gitea/routers/init.go:101 +0x333
code.gitea.io/gitea/routers.GlobalInit(0x3f1e980, 0xc00030ede0)
        /home/admin001/git/gitea/routers/init.go:125 +0x370
code.gitea.io/gitea/cmd.runWeb(0xc0003929a0, 0x0, 0x0)
        /home/admin001/git/gitea/cmd/web.go:116 +0x2c1
github.com/urfave/cli.HandleAction(0x1e3e220, 0x3a57a28, 0xc0003929a0, 0xc000d01d00, 0x0)
        /home/admin001/git/gitea/vendor/github.com/urfave/cli/app.go:490 +0xc8
github.com/urfave/cli.Command.Run(0x21ab091, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x22bee6d, 0x16, 0x0, ...)
        /home/admin001/git/gitea/vendor/github.com/urfave/cli/command.go:210 +0x9e8
github.com/urfave/cli.(*App).Run(0xc001476680, 0xc00003c180, 0x3, 0x3, 0x0, 0x0)
        /home/admin001/git/gitea/vendor/github.com/urfave/cli/app.go:255 +0x741
main.main()
        /home/admin001/git/gitea/main.go:109 +0x8a3
```

**Comment by vedranMv** [comment]
It runs after last merge of master, but it doesn't work as it used to. Only the person that created an issue can change the branch. I'm looking into it

**Comment by vedranMv** [comment]
Compiles and works as intended. What am I missing here for build to pass? :)

**Comment by techknowlogick** [comment]
@vedranMv `make fmt` will allow the file to pass lint

File: modules/notification/indexer/indexer.go
Comment (lunny) [inline modules/notification/indexer/indexer.go:153]: Since the branch name will not be indexed, I don't think we need to do the reindex.

Comment (vedranMv) [reply modules/notification/indexer/indexer.go:153]: Do I remove the reference to this function from all other files as well?
image

**Comment by TsakiDev** [comment]
Hi, will this get merged in 1.13.0?

That's would be really helpful instead of using ref #issue in the commits.

File: modules/notification/indexer/indexer.go
Comment (zeripath) [reply modules/notification/indexer/indexer.go:153]: I think it's better to keep it in. Just because we don't currently doesn't mean we won't in future.

File: services/issue/issue.go
Comment (zeripath) [inline services/issue/issue.go]: the named return is not really helpful here.

```suggestion
func ChangeIssueRef(issue *models.Issue, doer *models.User, ref string) error {
	oldRef := issue.Ref
	issue.Ref = ref

	if err := issue.ChangeRef(doer, oldRef); err != nil {
		return err
	}
```

File: templates/repo/issue/branch_selector_field.tmpl
Comment (zeripath) [inline templates/repo/issue/branch_selector_field.tmpl]: Should fix the indentation here

```suggestion
			</div>
		</div>
		<div id="branch-list" class="scrolling menu reference-list-menu">
			{{if .Issue.Ref}}
				<div class="item text small" data-id="" data-id-selector="#ref_selector"><strong><a href="#">{{$.i18n.Tr "repo.clear_ref"}}</a></strong></div>
			{{end}}	
			{{range .Branches}}
				<div class="item" data-id="refs/heads/{{.}}" data-name="{{.}}" data-id-selector="#ref_selector">{{.}}</div>
			{{end}}
		</div>
		<div id="tag-list" class="scrolling menu reference-list-menu" style="display: none">
			{{if .Issue.Ref}}
				<div class="item text small" data-id="" data-id-selector="#ref_selector"><strong><a href="#">{{.i18n.Tr "repo.clear_ref"}}</a></strong></div>
			{{end}}	
			{{range .Tags}}
				<div class="item" data-id="refs/tags/{{.}}" data-name="tags/{{.}}" data-id-selector="#ref_selector">{{.}}</div>
			{{end}}
```

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
After 8 years, this "branch selector" does nothing more than saving the branch/tag name into database and display it.

Are there real users who uses this feature? If no, could it be removed in new versions?

For anyone who still needs it, please provide feedbacks to  Prepare to remove the branch/tag selector in the issue sidebar #32744 

If no feedback, then this "branch/tag selector" will be removed in 1.23

**Pull request merged**
